tests/integration/helpers/lxd.py

```diff
@@ -0,0 +1,257 @@
+        """
+        return_code, stdout, _ = await run_in_unit(unit, "lxc list --format json")
+        if return_code != 0 or not stdout:
+            logger.error("Failed to run lxc list, %s", return_code)
+            return None
+        return json.loads(stdout)
+
+    async def is_desired_num_runners():
+        """Return whether there are desired number of lxc instances running.
+
+        Returns:
+            Whether the desired number of lxc runners have been reached.
+        """
+        lxc_instances = await get_lxc_instances()
+        if lxc_instances is None:
+            return False
+        return len(lxc_instances) == num
+
+    await wait_for(is_desired_num_runners, timeout=timeout, check_interval=30)
+
+    instances = await get_lxc_instances()
+    if not instances:
+        return
+
+    for instance in instances:
+        return_code, stdout, _ = await run_in_unit(unit, f"lxc exec {instance['name']} -- ps aux")
+        assert return_code == 0
+
+        assert stdout is not None
+        assert f"/bin/bash {Runner.runner_script}" in stdout
+
+
+async def run_in_lxd_instance(
+    unit: Unit,
+    name: str,
+    command: str,
+    env: dict[str, str] | None = None,
+    cwd: str | None = None,
+    timeout: int | None = None,
+) -> tuple[int, str | None, str | None]:
+    """Run command in LXD instance of a juju unit.
+
+    Args:
+        unit: Juju unit to execute the command in.
+        name: Name of LXD instance.
+        command: Command to execute.
+        env: Mapping of environment variable name to value.
+        cwd: Work directory of the command.
+        timeout: Amount of time to wait for the execution.
+
+    Returns:
+        Tuple of return code and stdout.
+    """
+    lxc_cmd = f"/snap/bin/lxc exec {name}"
+    if env:
+        for key, value in env.items():
+            lxc_cmd += f"--env {key}={value}"
+    if cwd:
+        lxc_cmd += f" --cwd {cwd}"
+    lxc_cmd += f" -- {command}"
+    return await run_in_unit(unit, lxc_cmd, timeout)
+
+
+async def start_test_http_server(unit: Unit, port: int):
+    """Start test http server.
+
+    Args:
+        unit: The unit to start the test server in.
+        port: Http server port.
+    """
+    await run_in_unit(
+        unit,
+        f"""cat <<EOT >> /etc/systemd/system/test-http-server.service
+[Unit]
+Description=Simple HTTP server for testing
+After=network.target
+
+[Service]
+User=ubuntu
+Group=www-data
+WorkingDirectory=/home/ubuntu
+ExecStart=python3 -m http.server {port}
+EOT""",
+    )
+    await run_in_unit(unit, "/usr/bin/systemctl daemon-reload")
+    await run_in_unit(unit, "/usr/bin/systemctl start test-http-server")
+
+    # Test the HTTP server
+    for _ in range(10):
+        return_code, stdout, _ = await run_in_unit(unit, f"curl http://localhost:{port}")
+        if return_code == 0 and stdout:
+            break
+        await sleep(3)
+    else:
+        assert False, "Timeout waiting for HTTP server to start up"
```

Comment:
I think we can leverage `helpers.wait_for` to make it a little more standardized when we need to wait for something :D